### PR TITLE
Watch for Rook third party resource before creating the cluster

### PR DESCRIPTION
**Pull request by travisn**
A single instance of rook is still supported and will always be created in the `rook` namespace. In the future we may add support for multiple instances. 

This change also updated the k8s client-go library to the 2.0 tag instead of using the 1.5 tag.

### CODE REVIEW

File: cmd/rook-operator/main.go [hunk 1]
```diff
@@ -75,7 +74,7 @@ func init() {
 	rootCmd.PersistentFlags().StringVar(&cfg.dataDir, "data-dir", "/var/lib/rook", "directory for storing configuration")
 	rootCmd.PersistentFlags().StringVar(&logLevelRaw, "log-level", "INFO", "logging level for logging/tracing output (valid values: CRITICAL,ERROR,WARNING,NOTICE,INFO,DEBUG,TRACE)")
 
-	rootCmd.Flags().BoolVar(&cfg.useAllDevices, "use-all-devices", false, "true to use all storage devices, false to require local device selection")
+	rootCmd.Flags().StringVar(&cfg.namespace, "namespace", "", "the namespace in which the operator is running (required)")
```
Comment (bassam) [inline cmd/rook-operator/main.go]: can the cluster name also come the TPR? or is the namespace name the cluster name?

File: demo/kubernetes/README.md [hunk 2]
```diff
@@ -14,12 +14,13 @@ Note that we are striving for even more smooth integration with Kubernetes in th
 
 ### Deploy Rook
 
-With your Kubernetes cluster running, Rook can be setup and deployed by simply deploying the [rook-operator](rook-operator.yaml) deployment manifest.
+With your Kubernetes cluster running, Rook can be setup and deployed by simply creating the [rook-operator](rook-operator.yaml) deployment and creating an [instance of rook](rook-operator-instance.yaml).
```
Comment (bassam) [inline demo/kubernetes/README.md]: I suggest you use "rook cluster" or "rook storage cluster" everywhere you're using "rook operator instance". There is a single operator, but zero or more clusters.

File: demo/kubernetes/rook-operator-instance.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+apiVersion: rook.io/v1beta1
+kind: Cluster
```
Comment (bassam) [inline demo/kubernetes/rook-operator-instance.yaml]: Can this be "RookCluster"?

File: demo/kubernetes/rook-operator.yaml [hunk 4]
```diff
@@ -12,5 +12,9 @@ spec:
...
       containers:
       - name: rook-operator
-        image: quay.io/rook/rook-operator:dev-2017-02-17-k8s
-        args: ["--container-version=dev-2017-02-17-k8s"]
+        image: quay.io/rook/rook-operator:dev-2017-02-23-k8s
+        env:
+        - name: ROOK_OPERATOR_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
```
Comment (bassam) [inline demo/kubernetes/rook-operator.yaml]: I'm not sure I follow this. So the operator can run in a namespace, and the cluster it creates can be in a different namespace?

File: demo/kubernetes/rook-operator-instance.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+metadata:
+  name: my-storage
+  namespace: rook
```
Comment (bassam) [inline demo/kubernetes/rook-operator-instance.yaml]: is this the namespace we will create MON and OSD pods in?

File: pkg/operator/cluster.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	clientset *kubernetes.Clientset
+	Metadata  v1.ObjectMeta `json:"metadata,omitempty"`
+	Spec      ClusterSpec   `json:"spec"`
+}
+
+type ClusterSpec struct {
+	// Version is the expected version of the rook container to run in the cluster.
+	// The rook-operator will eventually make the rook cluster version
+	// equal to the expected version.
+	Version string `json:"version"`
```
Comment (bassam) [inline pkg/operator/cluster.go]: Shouldn't namespace be here too?

File: pkg/operator/cluster.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+	// Start the OSDs
+	osds := osd.New(c.Namespace, c.Spec.Version, c.Spec.UseAllDevices)
+	err = osds.Start(c.clientset, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to start the osds. %+v", err)
+	}
+
+	// Start the object store
+	r := rgw.New(c.Namespace, c.Spec.Version, c.factory)
```
Comment (bassam) [inline pkg/operator/cluster.go]: why is this needed?

File: demo/kubernetes/rook-operator-instance.yaml (hunk 3 above)
Comment (travisn) [inline demo/kubernetes/rook-operator-instance.yaml]: Yes, looks like it's possible. i'm investigating.

Comment (travisn) [inline demo/kubernetes/rook-operator-instance.yaml]: Note that the kind is unique to the rook.io namespace. If the kind is `RookCluster`, the TPR will look a bit redundant like below. What if we call it `Storage` or keep it as `Cluster`?
```
NAME                   DESCRIPTION             VERSION(S)
rook-cluster.rook.io   Managed rook clusters   v1beta1
```

File: demo/kubernetes/README.md (hunk 2 above)
Comment (kokhang) [inline demo/kubernetes/README.md]: +1. The term Rook-operator-instance seems to implied that they are many operators

File: demo/kubernetes/rook-operator.yaml [hunk 8]
```diff
@@ -12,5 +12,9 @@ spec:
     spec:
       containers:
       - name: rook-operator
-        image: quay.io/rook/rook-operator:dev-2017-02-17-k8s
-        args: ["--container-version=dev-2017-02-17-k8s"]
+        image: quay.io/rook/rook-operator:dev-2017-02-23-k8s
```
Comment (kokhang) [inline demo/kubernetes/rook-operator.yaml]: I think we shouldnt be chaning the image tag. It should just be `image: quay.io/rook/rook-operator`. When developing, we can specify the tag. When we merge, we should publish a new 'latest' image.

File: demo/kubernetes/rook-operator.yaml (hunk 4 above)
Comment (kokhang) [inline demo/kubernetes/rook-operator.yaml]: im also not too clear about the namespace strategy. Could you quickly explain in which namespace you envision all the resources to be created? Including the rook-operator.

File: pkg/operator/event.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	Type   kwatch.EventType
+	Object *Cluster
+}
+
+type rawEvent struct {
+	Type   kwatch.EventType
+	Object json.RawMessage
+}
+
+func pollEvent(decoder *json.Decoder) (*Event, *unversioned.Status, error) {
```
Comment (kokhang) [inline pkg/operator/event.go]: Kubernetes already maintains a util class that does pooling, wait etc. We should use this instead rather than creating and maintaining our own. https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go

File: pkg/operator/operator.go [hunk 10]
```diff
@@ -12,140 +12,311 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+	defer func() {
+		for _, stop := range o.stopChMap {
+			close(stop)
+		}
+		o.waitCluster.Wait()
+	}()
+
+	eventCh, errCh := o.watch(watchVersion)
+
+	go func() {
```
Comment (kokhang) [inline pkg/operator/operator.go]: I probably dont understand much about how we are leveraging TPR but is there a need to watch it? Doesnt kubernetes handle that for us? If i create a Kind: RookCluster, the TPR will handle that. And if we need to configure it or updated it, we could leverage kubectl.

File: demo/kubernetes/rook-operator-instance.yaml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+apiVersion: rook.io/v1beta1
+kind: Cluster
+metadata:
+  name: my-storage
```
Comment (jbw976) [inline demo/kubernetes/rook-operator-instance.yaml]: how does the name field of 'my-storage' get used?

File: pkg/operator/cluster.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+type ClusterSpec struct {
+	// Version is the expected version of the rook container to run in the cluster.
+	// The rook-operator will eventually make the rook cluster version
+	// equal to the expected version.
+	Version string `json:"version"`
+
+	// Paused is to pause the control of the operator for the rook cluster.
+	Paused bool `json:"paused,omitempty"`
```
Comment (jbw976) [inline pkg/operator/cluster.go]: how would a k8s user toggle this field?  would this be in the `kubectl update` path?

File: demo/kubernetes/README.md (hunk 2 above)
Comment (travisn) [inline demo/kubernetes/README.md]: yes, renamed to rook cluster

File: pkg/operator/cluster_list.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+*/
+package operator
+
+import (
+	"encoding/json"
+
+	"k8s.io/client-go/pkg/api/unversioned"
+)
+
+// ClusterList is a list of etcd clusters.
```
Comment (jbw976) [inline pkg/operator/cluster_list.go]: etcd clusters?

File: demo/kubernetes/rook-operator.yaml (hunk 8 above)
Comment (travisn) [inline demo/kubernetes/rook-operator.yaml]: definitely agreed, we just have a bit of a chicken and egg. We have to publish the latest build that will support this. Right now the latest build is pre-kubernetes

File: demo/kubernetes/rook-operator.yaml (hunk 4 above)
Comment (travisn) [inline demo/kubernetes/rook-operator.yaml]: we need to talk through this

File: pkg/operator/cluster.go (hunk 6 above)
Comment (travisn) [inline pkg/operator/cluster.go]: yes, i'm thinking the namespace should be here

File: pkg/operator/event.go (hunk 9 above)
Comment (travisn) [inline pkg/operator/event.go]: we can actually only use packages in the client-go project, and i don't see wait helpers there. This is also what the etcd-operator is doing right now. Clearly we need this in client-go

File: pkg/operator/operator.go (hunk 10 above)
Comment (travisn) [inline pkg/operator/operator.go]: this is the extensibility point. We have to implement what happens when our TPR is created.

File: pkg/operator/operator.go [hunk 14]
```diff
@@ -12,140 +12,311 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+		pt := k8sutil.NewPanicTimer(time.Minute, "unexpected long blocking (> 1 Minute) when handling cluster event")
```
Comment (jbw976) [inline pkg/operator/operator.go]: is panicking the right thing we want to do here?  does the operator instance get scheduled and restarted somewhere else after the panic?

File: pkg/operator/operator.go [hunk 15]
```diff
@@ -12,140 +12,311 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+							break
+						}
+
+						// if anything has changed (or error on relist), we have to rebuild the state.
+						// go to recovery path
+						errCh <- ErrVersionOutdated
+						return
+					}
+
+					logger.Errorf("unexpected status response from API server: %v", st.Message)
```
Comment (jbw976) [inline pkg/operator/operator.go]: does this not need ot be sent along on the error channel?

File: pkg/operator/event.go (hunk 9 above)
Comment (kokhang) [inline pkg/operator/event.go]: I was referring to the polling util methods. Kuberenetes provides a bunch of utils files that are normally used by kubernetes dev use when developing services/apps on kubernetes. 

Here are some example of where they are used. https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/rbd/rbd.go#L26
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L43
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L37

Comment (kokhang) [inline pkg/operator/event.go]: Since this change is large enough, we can address this issue on a different PR. But i think it would be good to leverage what kubernetes already provides rather than us creating and maintaining our own :)

File: demo/kubernetes/rook-cluster.yaml [hunk 16]
```diff
@@ -0,0 +1,8 @@
+apiVersion: rook.io/v1beta1
+kind: Cluster
+metadata:
+  name: my-rook
+spec:
+  namespace: rook
+  version: dev-2017-02-24-k8s
+  useAllDevices: false
```
Comment (kokhang) [inline demo/kubernetes/rook-cluster.yaml]: nit: to add line end of the file just to get rid of the git warning.
